File: python/algorithms/math/number_theory/number_theory.py
"""数论：gcd、埃氏筛（模幂与扩欧见 fast_power / extended_gcd）。"""

from __future__ import annotations


def gcd(a: int, b: int) -> int:
    while b:
        a, b = b, a % b
    return abs(a)


def sieve(n: int) -> list[bool]:
    if n < 2:
        return [False] * (n + 1)
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    for i in range(2, int(n**0.5) + 1):
        if is_prime[i]:
            step = i
            start = i * i
            for j in range(start, n + 1, step):
                is_prime[j] = False
    return is_prime


if __name__ == "__main__":
    assert gcd(54, 24) == 6
    sp = sieve(30)
    assert sum(sp) == 10  # primes <=30 count
    assert len(sieve(0)) == 1 and not sieve(0)[0]
    assert sum(sieve(1)) == 0
    print("number_theory OK")

File: cpp/algorithms/math/number_theory/number_theory.cpp
// 数论：gcd、埃氏筛（模幂与扩欧见 fast_power / extended_gcd）
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

long long gcd_ll(long long a, long long b) {
    while (b) {
        long long t = a % b;
        a = b;
        b = t;
    }
    return abs(a);
}

vector<char> sieve(int n) {
    vector<char> is(n + 1, 1);
    if (n >= 0) is[0] = 0;
    if (n >= 1) is[1] = 0;
    for (int i = 2; i * i <= n; ++i)
        if (is[i])
            for (int j = i * i; j <= n; j += i) is[j] = 0;
    return is;
}

int main() {
    assert(gcd_ll(54, 24) == 6);
    auto sp = sieve(30);
    int c = accumulate(sp.begin(), sp.end(), 0);
    assert(c == 10);
    cout << "number_theory OK" << endl;
    return 0;
}
